eventsources/sources/kafka/start.go

```diff
@@ -72,6 +76,100 @@ func (el *EventListener) StartListening(ctx context.Context, dispatch func([]byt
 	log.Info("start kafka event source...")
 	kafkaEventSource := &el.KafkaEventSource
 
+	if kafkaEventSource.ConsumerGroup == nil {
+		return el.partitionConsumer(ctx, log, kafkaEventSource, dispatch)
+	} else {
+		return el.consumerGroupConsumer(ctx, log, kafkaEventSource, dispatch)
+	}
+}
+
+func (listener *EventListener) consumerGroupConsumer(ctx context.Context, log *zap.SugaredLogger, kafkaEventSource *v1alpha1.KafkaEventSource, dispatch func([]byte) error) error {
+	config := sarama.NewConfig()
+
+	version, err := sarama.ParseKafkaVersion(kafkaEventSource.ConsumerGroup.KafkaVersion)
+	if err != nil {
+		log.Errorf("Error parsing Kafka version: %v", err)
+		return err
+	}
+	config.Version = version
+
+	switch kafkaEventSource.ConsumerGroup.RebalanceStrategy {
+	case "sticky":
+		config.Consumer.Group.Rebalance.Strategy = sarama.BalanceStrategySticky
+	case "roundrobin":
+		config.Consumer.Group.Rebalance.Strategy = sarama.BalanceStrategyRoundRobin
+	case "range":
+		config.Consumer.Group.Rebalance.Strategy = sarama.BalanceStrategyRange
+	default:
+		log.Info("Invalid rebalance strategy using default: range")
+		config.Consumer.Group.Rebalance.Strategy = sarama.BalanceStrategyRange
+	}
+
+	if kafkaEventSource.TLS != nil {
+		tlsConfig, err := common.GetTLSConfig(kafkaEventSource.TLS.CACertPath, kafkaEventSource.TLS.ClientCertPath, kafkaEventSource.TLS.ClientKeyPath)
+		if err != nil {
+			return errors.Wrap(err, "failed to get the tls configuration")
+		}
+		config.Net.TLS.Config = tlsConfig
+		config.Net.TLS.Enable = true
+	}
+
+	consumer := Consumer{
+		ready:            make(chan bool),
+		dispatch:         dispatch,
+		logger:           log,
+		kafkaEventSource: kafkaEventSource,
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	client, err := sarama.NewConsumerGroup([]string{kafkaEventSource.URL}, kafkaEventSource.ConsumerGroup.GroupName, config)
+	if err != nil {
+		log.Errorf("Error creating consumer group client: %v", err)
+		cancel()
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			if err := client.Consume(ctx, []string{kafkaEventSource.Topic}, &consumer); err != nil {
```

Comment:
Is this a blocking call? for loop here means it keeps calling it, or do I miss anything?

Reply:
This function dose block.

```
Consume joins a cluster of consumers for a given list of topics and starts a blocking ConsumerGroupSession through the ConsumerGroupHandler.
The life-cycle of a session is represented by the following steps:
1. The consumers join the group (as explained in https://kafka.apache.org/documentation/#intro_consumers )
   and is assigned their "fair share" of partitions, aka 'claims'.
2. Before processing starts, the handler's Setup() hook is called to notify the user
   of the claims and allow any necessary preparation or alteration of state.
3. For each of the assigned claims the handler's ConsumeClaim() function is then called
   in a separate goroutine which requires it to be thread-safe. Any state must be carefully protected
   from concurrent reads/writes.
4. The session will persist until one of the ConsumeClaim() functions exits. This can be either when the
   parent context is cancelled or when a server-side rebalance cycle is initiated.
5. Once all the ConsumeClaim() loops have exited, the handler's Cleanup() hook is called
   to allow the user to perform any final tasks before a rebalance.
6. Finally, marked offsets are committed one last time before claims are released.
Please note, that once a rebalance is triggered, sessions must be completed within Config.Consumer.Group.Rebalance.Timeout. This means that ConsumeClaim() functions must exit as quickly as possible to allow time for Cleanup() and the final offset commit. If the timeout is exceeded, the consumer will be removed from the group by Kafka, which will cause offset commit failures.
```